src/reconstruction.hpp

```diff
@@ -313,21 +231,44 @@ double mp5_subcalc(double Fjm2, double Fjm1, double Fj, double Fjp1, double Fjp2
 }
 #undef MINMOD
 
-template <typename T, typename Tq>
-KOKKOS_INLINE_FUNCTION void MP5(const int d, const int n, const int k, const int j,
-                                const int i, const T &v, const Tq &ql, const Tq &qr) {
-  const int dir = d - 1;
-  const int di = (d == X1DIR ? 1 : 0);
-  const int dj = (d == X2DIR ? 1 : 0);
-  const int dk = (d == X3DIR ? 1 : 0);
-  ql(dir, n, k + dk, j + dj, i + di) =
-      mp5_subcalc(v(n, k - 2 * dk, j - 2 * dj, i - 2 * di), v(n, k - dk, j - dj, i - di),
-                  v(n, k, j, i), v(n, k + dk, j + dj, i + di),
-                  v(n, k + 2 * dk, j + 2 * dj, i + 2 * di));
-  qr(dir, n, k, j, i) =
-      mp5_subcalc(v(n, k + 2 * dk, j + 2 * dj, i + 2 * di), v(n, k + dk, j + dj, i + di),
-                  v(n, k, j, i), v(n, k - dk, j - dj, i - di),
-                  v(n, k - 2 * dk, j - 2 * dj, i - 2 * di));
+#pragma omp declare simd
+KOKKOS_INLINE_FUNCTION
+void MP5(const Real q0, const Real q1, const Real q2, const Real q3, const Real q4,
+         Real &ql, Real &qr) {
+  ql = mp5_subcalc(q0, q1, q2, q3, q4);
+  qr = mp5_subcalc(q4, q3, q2, q1, q0);
+}
+
+using WENO_LIKE = decltype(WENO5Z);
+template <WENO_LIKE F, typename T, typename U>
+KOKKOS_INLINE_FUNCTION
+void ReconLoop(parthenon::team_mbr_t &member, const int s, const int e,
+               T && pm2, T && pm1, U && p, T && pp1,
+               T && pp2, U && l, U && r) {
+  parthenon::par_for_inner(DEFAULT_INNER_LOOP_PATTERN, member, s, e,
+    [=](const int i) {
+      F(pm2[i], pm1[i], p[i], pp1[i], pp2[i], l[i], r[i]);
+    });
+}
+
+using LINEAR_LIKE = decltype(PiecewiseLinear);
+template <LINEAR_LIKE F, typename T, typename U>
+KOKKOS_INLINE_FUNCTION
+void ReconLoop(parthenon::team_mbr_t &member, const int s, const int e,
+               T && pm1, U && p, T && pp1, U && l, U && r) {
+  parthenon::par_for_inner(DEFAULT_INNER_LOOP_PATTERN, member, s, e, [=](const int i) {
+    F(pm1[i], p[i], pp1[i], l[i], r[i]);
+  });
+}
+
+using CONST_LIKE = decltype(PiecewiseConstant);
+template <CONST_LIKE F, typename T>
+KOKKOS_INLINE_FUNCTION
+void ReconLoop(parthenon::team_mbr_t &member, const int s, const int e,
+               T && p, T && l, T && r) {
+  parthenon::par_for_inner(DEFAULT_INNER_LOOP_PATTERN, member, s, e, [=](const int i) {
```

Comment:
I think it would make things more maintainable if all of the `ReconLoop` template overloads had the same call signature. Maybe I am missing something that prevents this?

Reply:
There's nothing that prevents this.  Actually my first version of this had one `ReconLoop` function where you pass in the `ReconType` enum and the switch that calls the right function was inside it.  Makes the code in fluid and radiation cleaner since it gets rid of those switch statements.  You could even try to make the `ReconLoop` function pull out the appropriate pointers so that nonsense is hidden away, though I'm not sure we're quite setup for that since the code in fluid and radiation are a bit different.  I went with the current approach because it works, didn't seem completely insane, and I argued that you shouldn't have to pass in 5 pointers for a 3 or 1 point stencil.